index.js

```diff
@@ -36,73 +262,238 @@ function cacheSet(key, value) {
  * @return {String} a hash representing the stats of this module and all its descendents
  */
 module.exports = function hashForDep(name, dir, _hashTreeOverride, _skipCache) {
-  var skipCache = false;
-  var key, hash;
+  var h;
+  
+  try {
+    var skipCache = (typeof _hashTreeOverride === 'function' || _skipCache === true) ? true : false;
 
-  if (typeof _hashTreeOverride === 'function' || _skipCache === true) {
-    skipCache = true;
-  } else {
-    key = cacheKey(name, dir);
-  }
+    setCaches(skipCache);
 
-  var heimdallNodeOptions = {
-    name: 'hashForDep(' + name + ')',
-    hashForDep: true,
-    dependencyName: name,
-    rootDir: dir,
-    skipCache: skipCache,
-    cacheKey: key
-  };
+    var hashTreeFn = (_hashTreeOverride || hashTree);
 
-  var heimdallNode = heimdall.start(heimdallNodeOptions, HashForDepSchema);
+    var nameDirKey = genNameDirKey(name, dir);
+    var realPathKey, realPath;
+    var hash;
 
-  if (skipCache === false && CACHE.has(key)) {
-    logger.info('cache hit: %s', key);
-    hash = CACHE.get(key);
-  } else {
-    var start = Date.now();
+    var heimdallNodeOptions = {
+      name: 'hashForDep(' + name + ')',
+      hashForDep: true,
+      dependencyName: name,
+      rootDir: dir,
+      skipCache: skipCache,
+      cacheKey: nameDirKey
+    };
+
+    h = heimdall.start(heimdallNodeOptions, HashForDepSchema);
+
+    var packageCacheEntry = computePackageInfo(name, dir, hashTreeFn);
 
-    var inputHashes = statPathsFor(name, dir).map(function(statPath) {
-      var hashFn = _hashTreeOverride || hashTree;
+    if (packageCacheEntry === NO_PACKAGE_ENTRY) {
+      // the name/dir pair does not resolve to a valid package.json file
+      return null;
+    }
+
+    if (packageCacheEntry.hash) {
+      return packageCacheEntry.hash;
+    }
 
-      heimdallNode.stats.paths++;
+    // We haven't computed the hash for this entry yet, though the
+    // packageCache entries are all present. It is possible that we
+    // have the hash for some dependencies, though.
 
-      var key = 'hashOf:' + statPath;
-      var hash;
+    // compute the full dependency list. There are no duplicates, because
+    // we check during gatherAllDependencies before inserting in the result.
+    // The keys of the returned object are the package baseDir values (the
+    // packageCacheEntry.rootDir). The values are the packageCacheEntry
+    // objects. 
+    var dependencies = Object.create({});
```

Comment:
I believe `dependencies` here is a growing dictionary right? If so, I think `Object.create(null)` would be more appropriate, as that immediately creates a dictionary, whereas `Object.create({})` first creates a new possible FastMap that inherits from `{}`, but since the result continues to grow, eventually the VM will bail out and fallback to a dictionary anyways. Which seems wasteful.

Reply:
Good point. I need to verify going to 'null', because there was a case (I don't think it was in this file, but I don't remember, it was a few weeks ago) where using 'null' meant that it didn't have access to a standard Object method that I was trying to use. The cache does grow, but only for enough time to get the dependencies for the original object transitively, then it's thrown out. Anyway, no objection to null, just let me verify that it actually doesn't trip over that other thing I found.

Reply:
Verified that using null works, so marking this resolved.